scheduler/scheduler-smartproxy-common/src/main/java/org/ow2/proactive/scheduler/smartproxy/common/AbstractSmartProxy.java

```diff
@@ -651,6 +638,55 @@ protected String prepareJobInput(Job job, String localInputFolder, String pushUR
         return push_url_updated;
     }
 
+    private String appendSlashIfNeeded(String url) {
+        if (url.endsWith("/")) {
+            return url;
+        }
+        return url + "/";
+    }
+
+    private void updateInputSpace(Job job, String inputFolder) throws NotConnectedException, PermissionException {
+        List<String> inputSpaces;
+        if (job.getInputSpace() != null) {
+            inputSpaces = Arrays.asList(Tools.dataSpaceConfigPropertyToUrls(job.getInputSpace()));
+        } else {
+            inputSpaces = getUserSpaceURIs();
+        }
+        StringBuilder inputSpace_url_updated = new StringBuilder();
+
+        for (String inputSpace : inputSpaces) {
+            if (inputSpace_url_updated.length() > 0) {
+                inputSpace_url_updated.append(" ");
+            }
+            inputSpace_url_updated.append(appendSlashIfNeeded(inputSpace));
+            inputSpace_url_updated.append(inputFolder);
+        }
+
+        job.setInputSpace(inputSpace_url_updated.toString());
+        log.debug("Input space of job " + job.getName() + " will be " + inputSpace_url_updated);
+    }
+
+    private void updateOutputSpace(Job job, String outputFolder) throws NotConnectedException, PermissionException {
+        List<String> outputSpaces;
+        if (job.getOutputSpace() != null) {
+            outputSpaces = Arrays.asList(Tools.dataSpaceConfigPropertyToUrls(job.getOutputSpace()));
+        } else {
+            outputSpaces = getUserSpaceURIs();
+        }
+        StringBuilder outputSpace_url_updated = new StringBuilder();
+
+        for (String inputSpace : outputSpaces) {
```

Comment:
this block is the same as the inputfiles....can we refactor and have only one method ?

Reply:
I can refactor the center part of both methods